### ISSUE TITLE
weaklifetime_par testsuite failure

### ISSUE THREAD

**Issue by ctk21** [ocamllabs/ocaml-multicore#588]
We have seen `weaklifetime_par` fail with a runtime assert failure in the Linux DEBUG CI runs:
<img width="744" alt="Screenshot 2021-06-16 at 12 54 59" src="https://user-images.githubusercontent.com/1682628/122214803-6c2f6000-cea2-11eb-8d63-99065d877429.png">
(full log [here](https://github.com/ocaml-multicore/ocaml-multicore/pull/586/checks?check_run_id=2838070591#step:5:5748))

I haven't had much luck recreating this locally.

**Comment by kayceesrk** [comment]
I can recreate it:

```
$ for i in {0..100}; do taskset --cpu-list 0,1 ./a.out; done
### OCaml runtime: debug mode ###
ok
### OCaml runtime: debug mode ###
ok
### OCaml runtime: debug mode ###
ok
### OCaml runtime: debug mode ###
ok
### OCaml runtime: debug mode ###
[00] file major_gc.c; line 941 ### Assertion failed: atomic_load_acq(&ephe_cycle_info.num_domains_todo) == atomic_load_acq(&ephe_cycle_info.num_domains_done)
Aborted (core dumped)
### OCaml runtime: debug mode ###
ok
### OCaml runtime: debug mode ###
```

**Comment by kayceesrk** [comment]
@Sudha247 I remember that you'd previously mentioned a similar assertion failure. Wondering if this is the same as the one that you might have previously seen.

**Comment by Sudha247** [comment]
So the assertion failure I mentioned is different and it is in #545. Although this one looks related to #545; `ephetest6` at times enters an infinite loop where it is not able to go to the next GC phase due to mismatch in ephemeron counters `ephe_cycle_info.num_domains_todo` and `ephe_cycle_info.num_domains_done`

**Comment by ctk21** [comment]
Thanks @kayceesrk, I can also recreate on a Linux box with `taskset --cpu-list 0,1` and `4.12+domains+effects` (`2ba91dfd0`).

**Comment by kayceesrk** [comment]
Fixed by #593.
